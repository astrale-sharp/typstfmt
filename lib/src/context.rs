use typst_syntax::LinkedNode;

use super::utils;
use crate::Config;

/// Writer is used to write your formatted output.
///
/// It comes with the following features :
/// - Markers: you place a mark by calling [Writer::mark], you can use this mark
/// to jump back and redo your formatting should you see it didn't respect some rules.
///
/// Example :
/// ```ignore
/// fn visit_params(/* */) {
///     let mark = self.mark();
///     visit_params_tight(self);
///     if !line_is_too_long(ctx.string_after_mark(mark)) {
///         // we're done yay
///         return
///     } else {
///         ctx.rewind(mark);
///         visit_params_long();
///     }
/// }
/// ```
/// - Indent, Dedent, Preserve: You must specify where you want indent to start and end.
/// It will be applied as a later step respecting the Preserve markers where indentation is
/// not applied (Raw, fmt::off, etc.)
pub(crate) struct Writer<'a> {
    pub(crate) config: Config,
    pub(crate) off: bool,
    pub(crate) buffer: &'a mut String,
}

pub(crate) struct Mark;

impl<'a> Writer<'a> {
    pub(crate) fn new(config: Config, buffer: &'a mut String) -> Self {
        Self {
            config,
            off: false,
            buffer,
        }
    }
    // needed when post process adds indentation level.
    pub fn mark_indent(&mut self) {
        // TODO
    }
    pub fn mark_dedent(&mut self) {
        // TODO
    }

    pub fn mark_preserve(&mut self) -> Mark {
        todo!()
    }

    pub fn wrap_preserve(&mut self, mark: Mark) {}

    pub(crate) fn push_node(&mut self, node: &LinkedNode) {
        self.buffer.push_str(node.text())
    }

    pub(crate) fn push_node_spaced(&mut self, n: &LinkedNode) {
        self.push_spaced(n.text())
    }

    pub(crate) fn push_str(&mut self, s: &str) {
        self.buffer.push_str(s)
    }

    pub(crate) fn push_spaced(&mut self, s: &str) {
        if s == "" {
            return;
        }

        self.buffer.push_str(s);

        if utils::last_line_length(self.buffer) >= self.config.max_line_length {
            self.new_line()
        } else {
            self.space()
        }
    }

    pub(crate) fn new_line(&mut self) {
        self.buffer.push('\n')
    }

    pub(crate) fn space(&mut self) {
        self.buffer.push(' ')
    }
}
