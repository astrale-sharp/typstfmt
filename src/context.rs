use super::*;

#[derive(Default)]
pub(crate) struct Ctx {
    pub(crate) config: Config,
    pub(crate) just_spaced: bool,
    pub(crate) consec_new_line: i32,
    pub(crate) off: bool,
}

/// you may push into your own buffer using this to ensure you push considering context
///
/// you may then push said buffer the final result.
impl Ctx {
    pub(crate) fn from_config(config: Config) -> Self {
        Self {
            config,
            ..Default::default()
        }
    }

    /// Pushes the string in the result avoiding:
    /// - putting two consecutive spaces.
    /// - putting more than two consecutive newlines.
    /// - trims the string if it DOES contain a newline.
    /// Won't work for indents.
    #[instrument(skip_all)]
    pub(crate) fn push_in(&mut self, s: &str, result: &mut String) {
        let s = if s.contains('\n') {
            s.trim_end_matches(' ')
        } else {
            s
        };
        for c in s.chars() {
            match c {
                ' ' => {
                    if self.just_spaced || result.ends_with(' ') {
                    } else {
                        self.just_spaced = true;
                        result.push(' ');
                    }
                }
                '\n' => {
                    if self.consec_new_line <= 1 {
                        self.consec_new_line += 1;
                        result.push('\n');
                    } else {
                        debug!("IGNORED newline");
                    }
                }
                _ => {
                    result.push(c);
                    self.lost_context();
                }
            }
        }
    }

    /// makes the context aware it missed info,
    /// should be called when pushing directly in result.
    pub(crate) fn push_raw_in(&mut self, s: &str, result: &mut String) {
        result.push_str(s);
        self.lost_context();
    }

    /// adds an indentation for each line the input except the first to match the current level of indentation.
    pub(crate) fn push_raw_indent(&mut self, s: &str, result: &mut String) {
        for (i, s) in s.split_inclusive('\n').enumerate() {
            if i == 0 {
                self.push_raw_in(s, result);
                continue;
            }
            let s = if s.ends_with("\n") {
                format!("{}\n", s.trim_end())
            } else {
                s.to_string()
            };
            self.push_raw_in(&format!("{}{}", self.get_indent(), s), result);
        }
    }

    // pub(crate) fn push_in_indent(&mut self, s: &str, result: &mut String) {
    //     let mut is_first = true;
    //     for s in s.lines() {
    //         let s = s.trim_end();
    //         if is_first {
    //             is_first = false;
    //             self.push_in(s, result);
    //             continue;
    //         }
    //         self.push_in("\n", result);
    //         self.push_in(
    //             format!("{}{}", self.get_indent(), s).trim_end_matches(' '),
    //             result,
    //         );
    //     }
    // }

    /// must be called when you cannot keep track of what you pushed
    /// so that context doesn't refuse your next pushes for no reasons.
    pub(crate) fn lost_context(&mut self) {
        self.just_spaced = false;
        self.consec_new_line = 0;
    }

    /// returns an indent using config to get it's length.
    pub(crate) fn get_indent(&self) -> String {
        if self.config.use_tabs {
            "\t".to_owned()
        } else {
            " ".repeat(self.config.indent_space)
        }
    }
}
