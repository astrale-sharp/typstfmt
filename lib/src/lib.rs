#![doc = include_str!("../README.md")]
#![warn(clippy::all)]

use itertools::Itertools;
use tracing::debug;
use tracing::instrument;
use tracing::warn;
use typst_syntax::ast::BinOp;
use typst_syntax::SyntaxKind;
use typst_syntax::SyntaxKind::*;
use typst_syntax::{parse, LinkedNode};
use Option::None;

mod config;
pub use config::Config;
mod context;
use context::Ctx;

mod utils;

mod args;
mod binary;
mod code_blocks;
mod markup;

#[must_use]
pub fn format(s: &str, config: Config) -> String {
    //replace tabs
    let s = &s.replace('\t', &" ".repeat(config.indent_space));

    let init = parse(s);
    let mut context = Ctx::from_config(config);
    let root = LinkedNode::new(&init);
    let s = visit(&root, &mut context);
    regex::Regex::new("( )+\n")
        .unwrap()
        .replace_all(&s, "\n")
        .to_string()
}

/// This is recursively called on the AST, the formatting is bottom up,
/// nodes will decide based on the size of their children and the max line length
/// how they will be formatted.
///
/// One assumed rule is that no kind should be formatting with surrounded space
#[instrument(skip_all,name = "V", fields(kind = format!("{:?}",node.kind())))]
fn visit(node: &LinkedNode, ctx: &mut Ctx) -> String {
    let mut res: Vec<String> = vec![];
    for child in node.children() {
        let child_fmt = visit(&child, ctx);
        res.push(child_fmt);
    }
    let res = match node.kind() {
        LineComment => format_comment_handling_disable(node, &res, ctx),
        _ if ctx.off => no_format(node, &res, ctx),
        Binary => binary::format_bin_left_assoc(node, &res, ctx),
        Named | Keyed => format_named_args(node, &res, ctx),
        ListItem | EnumItem | TermItem => format_list_enum(node, &res, ctx),
        CodeBlock => code_blocks::format_code_blocks(node, &res, ctx),
        Markup => markup::format_markup(node, &res, ctx),
        ContentBlock => markup::format_content_blocks(node, &res, ctx),
        Args | Params | Dict | Array | Destructuring | Parenthesized => {
            args::format_args(node, &res, ctx)
        }
        LetBinding => format_let_binding(node, &res, ctx),
        Conditional => conditional_format(node, &res, ctx),
        Raw | BlockComment => {
            ctx.lost_context();
            node.text().to_string()
        }
        _ => format_default(node, &res, ctx),
    };
    if node.children().count() == 0 {
        debug!("TOKEN : {:?}", node.kind());
    } else {
        debug!("PARENT: {:?}", node.kind());
    }
    res
}

/// formats a node for which no specific function was found. Last resort.
/// For the text of the node:
/// Trim spaces for Space nodes if they contain a linebreak.
/// avoids:
/// - putting two consecutive spaces.
/// - putting more than two consecutive newlines.
///
/// For the already formatted children, change nothing.
#[instrument(skip_all, ret)]
fn format_default(node: &LinkedNode, children: &[String], ctx: &mut Ctx) -> String {
    let mut res = String::new();
    ctx.push_in(node.text(), &mut res);
    for s in children {
        ctx.push_raw_in(s, &mut res);
    }
    res
}

fn no_format(parent: &LinkedNode, children: &[String], ctx: &mut Ctx) -> String {
    let mut res = String::new();
    ctx.push_raw_in(parent.text(), &mut res);
    for s in children {
        ctx.push_raw_in(s, &mut res);
    }
    res
}

fn deep_no_format(parent: &LinkedNode) -> String {
    let mut res: Vec<String> = vec![];
    for child in parent.children() {
        let child_fmt = deep_no_format(&child);
        res.push(child_fmt);
    }
    no_format(parent, &res, &mut Ctx::default())
}

fn conditional_format(parent: &LinkedNode, children: &[String], ctx: &mut Ctx) -> String {
    let mut res = String::new();
    ctx.push_raw_in(parent.text(), &mut res);
    for (s, node) in children.iter().zip(parent.children()) {
        match node.kind() {
            _ if ctx.off => res.push_str(node.text()),
            Space => {}
            If => {
                ctx.push_raw_in(s, &mut res);
                ctx.push_raw_in(" ", &mut res);
            }
            CodeBlock | ContentBlock => {
                ctx.push_raw_in(" ", &mut res);
                ctx.push_raw_in(s, &mut res);
            }
            Else => {
                ctx.push_raw_in(" ", &mut res);
                ctx.push_raw_in(s, &mut res);
                if node.next_sibling_kind() == Some(Conditional) {
                    ctx.push_raw_in(" ", &mut res);
                }
            }
            _ => ctx.push_raw_in(s, &mut res),
        }
    }
    res
}

#[instrument(skip_all, ret)]
pub(crate) fn format_named_args(parent: &LinkedNode, children: &[String], ctx: &mut Ctx) -> String {
    let mut res = String::new();
    for (s, node) in children.iter().zip(parent.children()) {
        match node.kind() {
            _ if ctx.off => res.push_str(node.text()),
            Show | Set => {
                ctx.push_raw_in(s, &mut res);
                ctx.push_in(" ", &mut res);
            }
            Colon => res.push_str(": "),
            Space => {}
            LineComment | BlockComment => ctx.push_raw_in(s, &mut res),
            _ => {
                ctx.push_raw_in(s, &mut res);
            }
        }
    }
    res
}

#[instrument(skip_all, ret)]
pub(crate) fn format_let_binding(
    parent: &LinkedNode,
    children: &[String],
    ctx: &mut Ctx,
) -> String {
    let mut res = String::new();
    for (s, node) in children.iter().zip(parent.children()) {
        match node.kind() {
            _ if ctx.off => res.push_str(&deep_no_format(&node)),
            Eq => {
                ctx.push_in(" ", &mut res);
                ctx.push_in(s, &mut res);
                ctx.push_in(" ", &mut res);
            }
            Space => ctx.push_in(s, &mut res),
            _ => {
                ctx.push_raw_in(s, &mut res);
            }
        }
    }
    res
}

fn format_comment_handling_disable(parent: &LinkedNode, _: &[String], ctx: &mut Ctx) -> String {
    ctx.lost_context();
    if parent.text().contains("typstfmt::off") {
        ctx.off = true;
    } else if parent.text().contains("typstfmt::on") {
        ctx.off = false;
    } else if parent.text().contains("typstfmt::") {
        warn!("your comment contains `typstfmt::` not followed by `on` or `off`, did you make a typo?");
    }
    parent.text().to_string()
}

fn format_list_enum(parent: &LinkedNode, children: &[String], ctx: &mut Ctx) -> String {
    let mut res = String::new();
    for (s, node) in children.iter().zip(parent.children()) {
        match node.kind() {
            _ if ctx.off => res.push_str(&deep_no_format(&node)),
            EnumMarker | ListMarker | TermMarker => {
                ctx.push_raw_in(node.text(), &mut res);
            }
            _ => {
                ctx.push_raw_indent(s, &mut res);
            }
        }
    }
    res
}

#[cfg(test)]
mod tests;
