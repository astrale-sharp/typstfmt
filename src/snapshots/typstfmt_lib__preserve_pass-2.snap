---
source: lib/src/lib.rs
expression: root
---
Markup: [WithSpacing(Destruct): Text("#"), FuncCall: [WithSpacing(StrongSpace): Text("f"), ParamsLike: [WithSpacing(WeakSpace): Text("("), ContentBlock: [WithSpacing(WeakSpace): Text("["), Markup: [WithSpacing(WeakSpace): Text("notp")], WithSpacing(WeakSpace): Text("]")], WithSpacing(StrongSpace): Text(","), Space: Text(" "), Preserve(1): Text("// typstfmt::off"), Preserve(1): Text("\n"), Preserve(1): Text(")")]], Preserve(1): Text("\n"), Preserve(2): Text("// typstfmt::off"), Preserve(2): Text("\n"), Preserve(1): Text("// typstfmt::on"), Preserve(1): Text("\n"), Preserve(1): Text("p"), Preserve(1): Text("\n"), Comment: Text("// typstfmt::on"), Space: Text("\n"), WithSpacing(WeakSpace): Text("notp")]
